doc/internal/consensus/Consensus-Protocol.md

```diff
@@ -0,0 +1,80 @@
+ 
+
+#### Implementation
+
+* A separate `interal/raft` folder will have a raft implementation which provides APIs for each server to call.
+* 
+### Log Replication
+
+#### Spec
+
+* Pre-log replication: Once a leader is elected, it starts servicing the client. The leader appends a new request to its `New Entry` log then issues `AppendEntriesRPC` in parallel to all its peers. 
+* Successful log: When all logs have been applied successfully to all follower machines, the leader applies the entry to its state machine and returns the result to the client.
+* Repeating `AppendEntries`: `AppendEntriesRPC` are repeated indefinitely until all followers eventually store all log entries.
+* Log entry storage: Log entries are a queue of state machine commands which are applied to that particular state machine. Log entries are associated with a term number to indicate the term of application of that log along with an integer index to identify a particular logs position.
+* Committed entry: A log entry is called committed once its replicated on the majority of the servers in the cluster. Once an entry is committed, it commits all the previous entries in the leaders log, including the entries created by the previous leaders.
+* The  leader keeps track of the highest known index that it knows is committed and it is included in all the future `AppendEntriesRPC` (including heartbeats) to inform other servers.
+* Theres some issue about log committing - paragraph says its committed when its applied everywhere and also says its applied everywhere once its committed. 
+* Log matching property: 
+  * If two entries in different logs have the same index and term, then they store the same com-mand.
+  * If two entries in different logs have the same index and term, then the logs are identical in allpreceding entries.
+ * 
+* Add more from section 3.6
+
+#### Implementation
+
+### Safety
+
+## A strict testing mechanism
+
+The testing mechanism to be implemented will enable us in figuring out the problems existing in the implementation leading to a more resilient system.
+We have to test for the following basic failures:
+1. Network partitioning.
+2. Un-responsive peers.
+3. Overloading peer.
+4. Corruption of data in transit.
+
+## Graceful handling of failures
+
+Accepting failures exist and handling them gracefully enables creation of more resilient and stable systems. Having _circuit breakers_, _backoff mechanisms in clients_ and _validation and coordination mechanisms_ are some of the pointers to be followed. 
+
+## Running Lbadd on Raft
+
+* Background: Raft is just a consensus protocol that helps keep different database servers in sync. We need methods to issue a command and enable the sync between servers.
+* Logistics: The `AppendEntriesRPC` will have the command to be executed by the client. This command goes through the leader, is applied by all the followers and then committed by the leader. Thus ensuring an in-sync distributed database.
```

Comment:
This command is invoked by leader. Each client request contains a command to be executed by the replicated state machines. The leader append that command to its log entries and then issues AppendEntriesRPC to rest of the follower. It is also use as heartbeat by the leader

So command doesn't go through the leader, is invoked and also not executed by the client

Reply:
Everything goes through the leader and no-one else.
The log in `AppendEntries` RPC will have a compiled version of the SQL stmt that the follower nodes will apply.